cli/src/ssh_agent.rs

```diff
@@ -0,0 +1,304 @@
+    Ok(())
+}
+
+async fn list_keys(
+    ssh_agent_sock: &mut tokio::net::UnixStream,
+) -> Result<Vec<PublicKey>, ListKeysError> {
+    send_msg(
+        ssh_agent_sock,
+        Message {
+            type_: SSH_AGENTC_REQUEST_IDENTITIES,
+            contents: Vec::new(),
+        },
+    )
+    .await?;
+    let reply = read_msg(ssh_agent_sock).await?;
+    if reply.type_ != SSH_AGENT_IDENTITIES_ANSWER {
+        return Err(ListKeysError::UnexpectedAnswer(reply.type_));
+    }
+    let mut reader = sshkeys::Reader::new(&reply.contents);
+    let nkeys = usize::try_from(reader.read_u32()?)?;
+    let mut keys = Vec::with_capacity(nkeys);
+    let mut remaining = nkeys;
+    while remaining > 0 {
+        let key_blob = reader.read_bytes()?;
+        let comment = reader.read_string()?;
+        let mut key = PublicKey::from_bytes(&key_blob)?;
+        key.comment = Some(comment);
+        keys.push(key);
+        remaining -= 1;
+    }
+
+    Ok(keys)
+}
+
+/// Return the first SSH public key that matches a given JWK
+fn select_key(keys: Vec<PublicKey>, jwk: Option<&JWK>) -> Result<(JWK, PublicKey), SignError> {
+    let jwk = match jwk {
+        Some(jwk) => jwk,
+        None => {
+            if keys.len() > 1 {
+                return Err(SignError::TooManyKeys);
+            }
+            let pk = keys.into_iter().next().ok_or(SignError::NoKeys)?;
+            let jwk = ssi::ssh::ssh_pkk_to_jwk(&pk.kind)?;
+            return Ok((jwk, pk));
+        }
+    };
+    let thumbprint = jwk
+        .thumbprint()
+        .map_err(|e| SignError::JWKThumbprint(e.to_string()))?;
+    for pk in keys {
+        let sshkey_jwk = ssi::ssh::ssh_pkk_to_jwk(&pk.kind)?;
+        let sshkey_thumbprint = sshkey_jwk
+            .thumbprint()
+            .map_err(|e| SignError::JWKThumbprint(e.to_string()))?;
+        if sshkey_thumbprint == thumbprint {
+            return Ok((sshkey_jwk, pk));
+        }
+    }
+    Err(SignError::NoKeys)
+}
+
+pub async fn sign_with_pk(
+    document: &(dyn ssi::ldp::LinkedDataDocument + Sync),
+    options: LinkedDataProofOptions,
+    jwk_opt: Option<&JWK>,
+    ssh_agent_sock: &mut tokio::net::UnixStream,
+) -> Result<ssi::vc::Proof, SignError> {
+    let keys = list_keys(ssh_agent_sock).await?;
+    let (jwk, pk) = select_key(keys, jwk_opt)?;
+    let prep = LinkedDataProofs::prepare(document, &options, &jwk).await?;
+    let mut signing_input_bytes = match prep.signing_input {
+        ssi::ldp::SigningInput::Bytes(ref bytes) => bytes.0.to_vec(),
+        _ => Err(SignError::UnsupportedSigningInputFormat)?,
+    };
+    // TODO: check key size is 2048 bits or larger.
+    // https://datatracker.ietf.org/doc/html/rfc7518#section-3.3
+    let mut writer = sshkeys::Writer::new();
+    writer.write_bytes(&pk.encode());
+    if let Some(ref jws_header) = prep.jws_header {
+        if jws_header.algorithm == Algorithm::ES256 {
+            // TODO: confirm with another implementation that this is correct
```

Comment:
It's not clear from the specs (or even code, I had a quick look) whether ssh agent is supposed to do the hashing itself. So +1  for that TODO :grimacing:

Reply:
I confirmed ES256 with other implementations in https://github.com/spruceid/didkit/pull/166 - that was after removing the hashing from ssi, in https://github.com/spruceid/ssi/pull/201/commits/d07bf6d3cb0b5b3133477afa05edcb4cc2c4a44f. It should be easier to address this TODO after JWT VC issuance is merged.

Reply:
The hashing was not needed there. Removed in e06c69c